File: S2_KP/Lectures/codes07/TablicaJavaDoc/TablicaBool.java
/**
 *  Klasa definiujaca wyjatek wykorzystywany w klasie {@link TablicaBool TablicaBool}
 */
class MojException extends Exception {
  /** Konstruktor podstawowy 
   *  @param w komunikat b³êdu
   */
  MojException(String w) { super(w); };
}

/** 
 *  Klasa definiuj¹ca tablice boolowska
 *  @version 1.0
 *  @author Wojciech Macyna */
public class TablicaBool {

  /** Tablica przechowujaca wartosci */
  private boolean[] t;

  /** Funkcja zwracajaca wartosci elementu tablicy 
   *  @param i indeks zwracanego elementu
   *  @return wartosc zwracanego elementu
   *  @throws MojException wyjatek w przypadku blednych indeksow
   *  @see TablicaBool
   *  @see TablicaBool
    */
  public boolean getValue( int i ) throws MojException
  {
    if( (i<0) || (i>=t.length) ) throw new MojException("Przekroczenie zakresu");
    return t[i];
    /// 
  }

  /** Funkcja ustawiajaca wartosc elementu tablicy 
   *  @param i indeks ustawianego elementu
   *  @param b wartosc ustawianego elementu
   *  @throws MojException wyjatek w przypadku blednych indeksow
   * @see TablicaBool#getValue(int) 
   */
  public void setValue( int i, boolean b ) throws MojException
  {
    if( (i<0) || (i>=t.length) ) throw new MojException("Przekroczenie zakresu");
    t[i] = b;
  }

  /** Podstawowy konstruktor
   *  @param i deklarowana wielkosc tablicy
   *  @throws MojException wyjatek w przypadku blednego rozmiaru 
   */
  TablicaBool( int i ) throws MojException
  {
     if(i<=0) throw new MojException("Za mala wielkosc danych");
     try {
       t = new boolean[i];
     }
     catch( OutOfMemoryError e ) {
       throw new MojException("Za malo pamieci");
     }
  }

/**
 * Glowna funkcja testujaca klase
 * @param arg 
 */
  public static void main( String[] arg ) {
    TablicaBool t;
    int n = 1<<4;

    try {
      t = new TablicaBool(n);
      t.setValue(20,true);
    }
    catch( Exception e ) {
      System.out.println("2 - "+ e.getMessage() );
    }
    catch( Error e ) {
      System.out.println("3 - "+ e.getMessage() );
    }
  }
}

File: S2_KP/Lectures/codes07/TablicaCppDoxygen/TablicaBool.cpp
#include<string>
#include<exception>

#include "TablicaBool.h"

using namespace std;

//******************************************************************************

MojException::MojException(string ss) : s(ss) {}

MojException::~MojException() throw() {}

const char*MojException::what() const throw() { return s.c_str(); }

//******************************************************************************

bool TablicaBool::getValue( long i ) throw (MojException) {
  if( (i<0) || (i>=n) ) throw MojException("Zly indeks tablicy");
  return t[i];
}

void TablicaBool::setValue( long i, bool b ) throw (MojException) {
  if( (i<0) || (i>=n) ) throw MojException("Zly indeks tablicy");
  t[i] = b;
}

TablicaBool::TablicaBool( long n ) throw (MojException) {
  if(n<=0) throw MojException("Zly rozmiar tablicy");
  try {
    t = new bool[n];
  }
  catch(bad_alloc& w) {
    throw MojException("Za malo pamieci");
  }
  this->n=n;
}

TablicaBool::~TablicaBool() { delete [] t; }

